crates/ty_python_semantic/src/types/infer.rs

```diff
@@ -8140,7 +8140,36 @@ impl<'db, 'ast> TypeInferenceBuilder<'db, 'ast> {
                     slice_ty,
                 )
             }
+            // If the value type is a union make sure to union the load types.
+            // For example:
+            // val: tuple[int] | tuple[str]
+            // val[0] can be an int or str type
+            (Type::Union(union_ty), _, _) => {
+                let value_types = union_ty.elements(self.db());
+                let loads_types: Vec<Type<'db>> = value_types
+                    .iter()
+                    .copied()
+                    .map(|ty| self.infer_subscript_expression_types(value_node, ty, slice_ty))
+                    .collect();
+
+                UnionType::from_elements(self.db(), loads_types)
+            }
+            (Type::Intersection(intersection_ty), _, _) => {
+                let value_types = intersection_ty.positive(self.db());
+
+                let mut intersection_builder: IntersectionBuilder<'db> =
+                    IntersectionBuilder::new(self.db());
 
+                value_types
+                    .iter()
+                    .copied()
+                    .map(|ty| self.infer_subscript_expression_types(value_node, ty, slice_ty))
+                    .for_each(|ty| {
+                        intersection_builder = intersection_builder.clone().add_positive(ty);
+                    });
+
+                intersection_builder.build()
```

Comment:
The `clone` here isn't needed; we want to consume/replace the builder each time through. There's also a `fold` pattern that we've [used elsewhere](https://github.com/astral-sh/ty/issues/625) that will simplify this a bit:

```suggestion
                value_types
                    .iter()
                    .copied()
                    .map(|ty| self.infer_subscript_expression_types(value_node, ty, slice_ty))
                    .fold(
                        IntersectionBuilder::new(db),
                        IntersectionBuilder::add_positive,
                    )
                    .build()
```